File: py/src/draw_shape.py
"""
draw_shape.py — Atomic shape-drawing function

Draws a single shape as a matplotlib patch. Knows nothing about trials, the
Ebbinghaus illusion, or experimental design.
"""

import matplotlib.patches as mpatches
import numpy as np


def draw_shape(
    ax,
    shape: str,
    x: float,
    y: float,
    size: float,
    color: str = "black",
    fill=None,
    alpha: float = 1.0,
):
    """Draw a single shape on a matplotlib Axes.

    Args:
        ax: matplotlib Axes to draw on.
        shape: "circle" or "square".
        x: Center x-coordinate.
        y: Center y-coordinate.
        size: Radius (circle) or half side-length (square), in plot units.
        color: Border/stroke color.
        fill: Fill color. None = no fill (transparent).
        alpha: Opacity (0-1).
    """
    assert shape in ("circle", "square"), f"Unknown shape: {shape}"
    assert size > 0

    facecolor = fill if fill is not None else "none"

    if shape == "circle":
        patch = mpatches.Circle(
            (x, y), radius=size,
            edgecolor=color, facecolor=facecolor, alpha=alpha, linewidth=1,
        )
    elif shape == "square":
        patch = mpatches.Rectangle(
            (x - size, y - size), width=2 * size, height=2 * size,
            edgecolor=color, facecolor=facecolor, alpha=alpha, linewidth=1,
        )

    ax.add_patch(patch)

File: R/App/src/draw_shape.R
# =============================================================================
# draw_shape.R — Atomic shape-drawing function
# =============================================================================
# Draws a single shape as a ggplot layer. Knows nothing about trials, the
# Ebbinghaus illusion, or experimental design.
#
# Returns a list of ggplot geoms that can be added to a plot with `+`.
# =============================================================================

library(ggplot2)
library(ggforce)

#' Draw a single shape as ggplot layer(s)
#'
#' @param shape Character: "circle" or "square"
#' @param x Numeric: center x-coordinate
#' @param y Numeric: center y-coordinate
#' @param size Numeric: radius (circle) or half side-length (square), in plot units
#' @param color Character: border/stroke color
#' @param fill Character or NA: fill color. NA = no fill (transparent).
#' @param alpha Numeric [0, 1]: opacity. Default 1.
#'
#' @return A list of ggplot2 geom layers
draw_shape <- function(shape, x, y, size, color = "black", fill = NA,
                       alpha = 1) {

  # --- Validate inputs ---
  stopifnot(
    is.character(shape), length(shape) == 1,
    shape %in% c("circle", "square"),
    is.numeric(x), length(x) == 1,
    is.numeric(y), length(y) == 1,
    is.numeric(size), length(size) == 1, size > 0,
    is.character(color), length(color) == 1,
    length(fill) == 1,
    is.numeric(alpha), length(alpha) == 1, alpha >= 0, alpha <= 1
  )

  # Resolve fill: NA means transparent
  fill_value <- if (is.na(fill)) NA_character_ else fill

  if (shape == "circle") {
    layer <- ggforce::geom_circle(
      aes(x0 = .env$x, y0 = .env$y, r = .env$size),
      color = color,
      fill  = fill_value,
      alpha = alpha
    )
  } else if (shape == "square") {
    # Square: size = half side-length, so side = 2 * size
    half <- size
    layer <- annotate(
      "rect",
      xmin  = x - half,
      xmax  = x + half,
      ymin  = y - half,
      ymax  = y + half,
      color = color,
      fill  = fill_value,
      alpha = alpha
    )
  }

  layer
}
